File: source_py/ebook_renamer/jsonoutput.py
"""
JSON output functionality for the ebook renamer.
"""

import json
import os
from typing import List, Dict, Any

from .types import (
    OperationsOutput, RenameOperation, DuplicateGroup, 
    DeleteOperation, TodoItem, FileInfo
)


class JSONOutput:
    """Handles JSON output generation with deterministic sorting."""
    
    @staticmethod
    def from_results(clean_files: List[FileInfo], 
                    duplicate_groups: List[List[str]], 
                    files_to_delete: List[str], 
                    todo_items: List[Dict[str, str]], 
                    target_dir: str) -> OperationsOutput:
        """Create an OperationsOutput from processing results."""
        output = OperationsOutput(
            renames=[],
            duplicate_deletes=[],
            small_or_corrupted_deletes=[],
            todo_items=[]
        )
        
        # Add renames
        renames = []
        for file in clean_files:
            if file.new_name is not None:
                from_path = JSONOutput._make_relative_path(file.original_path, target_dir)
                to_path = JSONOutput._make_relative_path(file.new_path, target_dir)
                
                renames.append(RenameOperation(
                    from_path=from_path,
                    to_path=to_path,
                    reason="normalized"
                ))
        
        # Sort renames by 'from' path for deterministic output
        renames.sort(key=lambda x: x.from_path)
        output.renames = renames
        
        # Add duplicate deletions
        duplicate_deletes = []
        for group in duplicate_groups:
            if len(group) > 1:
                keep_path = JSONOutput._make_relative_path(group[0], target_dir)
                delete_paths = [JSONOutput._make_relative_path(path, target_dir) 
                              for path in group[1:]]
                # Sort delete paths for deterministic output
                delete_paths.sort()
                
                duplicate_deletes.append(DuplicateGroup(
                    keep=keep_path,
                    delete=delete_paths
                ))
        
        # Sort duplicate groups by 'keep' path for deterministic output
        duplicate_deletes.sort(key=lambda x: x.keep)
        output.duplicate_deletes = duplicate_deletes
        
        # Add small/corrupted deletions
        small_deletes = []
        for path in files_to_delete:
            small_deletes.append(DeleteOperation(
                path=JSONOutput._make_relative_path(path, target_dir),
                issue="deleted"
            ))
        
        # Sort by path for deterministic output
        small_deletes.sort(key=lambda x: x.path)
        output.small_or_corrupted_deletes = small_deletes
        
        # Add todo items (already sorted by category and file in CLI)
        output.todo_items = [TodoItem(**item) for item in todo_items]
        
        return output
    
    @staticmethod
    def to_json(output: OperationsOutput) -> str:
        """Convert the OperationsOutput to a JSON string."""
        # Convert to dict for JSON serialization
        data = {
            "renames": [
                {
                    "from": op.from_path,
                    "to": op.to_path,
                    "reason": op.reason
                }
                for op in output.renames
            ],
            "duplicate_deletes": [
                {
                    "keep": group.keep,
                    "delete": group.delete
                }
                for group in output.duplicate_deletes
            ],
            "small_or_corrupted_deletes": [
                {
                    "path": op.path,
                    "issue": op.issue
                }
                for op in output.small_or_corrupted_deletes
            ],
            "todo_items": [
                {
                    "category": item.category,
                    "file": item.file,
                    "message": item.message
                }
                for item in output.todo_items
            ]
        }
        
        return json.dumps(data, indent=2, ensure_ascii=False)
    
    @staticmethod
    def _make_relative_path(path: str, target_dir: str) -> str:
        """Convert an absolute path to a relative path using forward slashes."""
        # Convert to relative path
        try:
            rel_path = os.path.relpath(path, target_dir)
        except ValueError:
            # Fallback to absolute path if relative conversion fails
            rel_path = path
        
        # Convert to forward slashes for JSON output (POSIX-style)
        rel_path = rel_path.replace('\\', '/')
        
        # Handle case where path is the same as target directory
        if rel_path == '.':
            rel_path = ''
        
        return rel_path

File: source_go/internal/jsonoutput/jsonoutput.go
package jsonoutput

import (
	"encoding/json"
	"fmt"
	"path/filepath"
	"sort"
	"strings"

	"github.com/ebook-renamer/go/internal/types"
)

// FromResults creates an OperationsOutput from processing results
func FromResults(cleanFiles []*types.FileInfo, duplicateGroups [][]string, filesToDelete []string, todoItems []types.TodoItem, targetDir string) (*types.OperationsOutput, error) {
	output := &types.OperationsOutput{
		Renames:                 []types.RenameOperation{},
		DuplicateDeletes:        []types.DuplicateGroup{},
		SmallOrCorruptedDeletes: []types.DeleteOperation{},
		TodoItems:               []types.TodoItem{},
	}

	// Add renames
	renames := []types.RenameOperation{}
	for _, file := range cleanFiles {
		if file.NewName != nil {
			fromPath := makeRelativePath(file.OriginalPath, targetDir)
			toPath := makeRelativePath(file.NewPath, targetDir)

			renames = append(renames, types.RenameOperation{
				From:   fromPath,
				To:     toPath,
				Reason: "normalized",
			})
		}
	}
	// Sort renames by 'from' path for deterministic output
	sort.Slice(renames, func(i, j int) bool {
		return renames[i].From < renames[j].From
	})
	output.Renames = renames

	// Add duplicate deletions
	duplicateDeletes := []types.DuplicateGroup{}
	for _, group := range duplicateGroups {
		if len(group) > 1 {
			keepPath := makeRelativePath(group[0], targetDir)
			var deletePaths []string
			for _, path := range group[1:] {
				deletePaths = append(deletePaths, makeRelativePath(path, targetDir))
			}
			// Sort delete paths for deterministic output
			sort.Strings(deletePaths)

			duplicateDeletes = append(duplicateDeletes, types.DuplicateGroup{
				Keep:   keepPath,
				Delete: deletePaths,
			})
		}
	}
	// Sort duplicate groups by 'keep' path for deterministic output
	sort.Slice(duplicateDeletes, func(i, j int) bool {
		return duplicateDeletes[i].Keep < duplicateDeletes[j].Keep
	})
	output.DuplicateDeletes = duplicateDeletes

	// Add small/corrupted deletions
	smallDeletes := []types.DeleteOperation{}
	for _, path := range filesToDelete {
		smallDeletes = append(smallDeletes, types.DeleteOperation{
			Path:  makeRelativePath(path, targetDir),
			Issue: "deleted",
		})
	}
	// Sort by path for deterministic output
	sort.Slice(smallDeletes, func(i, j int) bool {
		return smallDeletes[i].Path < smallDeletes[j].Path
	})
	output.SmallOrCorruptedDeletes = smallDeletes

	// Add todo items (already sorted by category and file in CLI)
	output.TodoItems = todoItems

	return output, nil
}

// ToJSON converts the OperationsOutput to a JSON string
func ToJSON(output *types.OperationsOutput) (string, error) {
	jsonBytes, err := json.MarshalIndent(output, "", "  ")
	if err != nil {
		return "", fmt.Errorf("JSON serialization failed: %w", err)
	}
	return string(jsonBytes), nil
}

// makeRelativePath converts an absolute path to a relative path using forward slashes
func makeRelativePath(path, targetDir string) string {
	// Convert to relative path
	relPath, err := filepath.Rel(targetDir, path)
	if err != nil {
		// Fallback to absolute path if relative conversion fails
		relPath = path
	}

	// Convert to forward slashes for JSON output (POSIX-style)
	relPath = strings.ReplaceAll(relPath, "\\", "/")

	// Handle case where path is the same as target directory
	if relPath == "." {
		relPath = ""
	}

	return relPath
}
